### PR TITLE
Added wavelength correction based on cross-correlation with telluric spectrum

### PR DESCRIPTION
**Pull request by ricolandman**

### CODE REVIEW

File: pycrires/pipeline.py [hunk 1]
```diff
@@ -3726,108 +3726,90 @@ def correct_wavelengths(self, nod_ab: str = "A") -> None:
...
-        print(" [DONE]")
-
-        # Objective function for maximizing the correlation
-        # with the telluric model spectrum from SkyCalc
-
-        def corr_func(param, *args):
-            obs_spec, model_interp = args[0], args[1]
-            wavel_new = param[0] + param[1] * obs_spec[:, 0]  # (nm)
+        fits_files = list(self.file_dict[f'OBS_NODDING_EXTRACT{nod_ab}'].keys())
+        #print(files)
```
Comment (tomasstolker) [inline pycrires/pipeline.py:3739]: The commented line can be removed

File: pycrires/pipeline.py [hunk 2]
```diff
@@ -3726,108 +3726,90 @@ def correct_wavelengths(self, nod_ab: str = "A") -> None:
...
-            #     mode="valid",
-            #     method="direct",
-            # )
-
-            return corr_value
-
-        transm_interp = interpolate.interp1d(
-            transm_spec[:, 0], transm_spec[:, 1], kind="linear", bounds_error=True
-        )
+        #fits_file = f"{self.path}/product/cr2res_obs_nodding_extracted{nod_ab}_000.fits"
```
Comment (tomasstolker) [inline pycrires/pipeline.py:3741]: These commented lines can also be removed

File: pycrires/pipeline.py [hunk 3]
```diff
@@ -3726,108 +3726,90 @@ def correct_wavelengths(self, nod_ab: str = "A") -> None:
...
+                # Loop over spectral orders
+                for spec_name in spec_orders:
+                    # Extract WL, SPEC, and ERR for given order/detector
+                    wavel = hdu_list[f"CHIP{i_det+1}.INT1"].data[spec_name + "_WL"]
+                    spec = hdu_list[f"CHIP{i_det+1}.INT1"].data[spec_name + "_SPEC"]
+                    err = hdu_list[f"CHIP{i_det+1}.INT1"].data[spec_name + "_ERR"]
+
+                    #Remove continuum and nans of spectra
+                    nans = np.isnan(spec)
+                    continuum = ndimage.gaussian_filter1d(spec[~nans], 100, mode='reflect')
```
Comment (tomasstolker) [inline pycrires/pipeline.py:3770]: Should we make the width of the kernel a free parameter and set the default value to 100? I would also change the `mode` to 'nearest'.

File: pycrires/pipeline.py [hunk 4]
```diff
@@ -3726,108 +3726,90 @@ def correct_wavelengths(self, nod_ab: str = "A") -> None:
...
+                    spec_flat = spec[~nans]/continuum - 1
```
Comment (tomasstolker) [inline pycrires/pipeline.py:3771]: Why is the -1 needed? Maybe change to -1. such that it is clear that `spec_flat` is a float (although it works fine like this).

File: pycrires/pipeline.py [hunk 5]
```diff
@@ -3726,108 +3726,90 @@ def correct_wavelengths(self, nod_ab: str = "A") -> None:
...
+
+                    #Don't use the edges as that sometimes gives problems
+                    spec_flat = spec_flat[20:-20]
```
Comment (tomasstolker) [inline pycrires/pipeline.py:3774]: The amount of pixels that is cropped could also be a parameter of the function?

File: pycrires/pipeline.py [hunk 6]
```diff
@@ -3726,108 +3726,90 @@ def correct_wavelengths(self, nod_ab: str = "A") -> None:
...
+                    #Prepare cross-correlation grid
+                    a_grid = np.linspace(0.9, 1.1, 100)[:,np.newaxis, np.newaxis]
+                    b_grid = np.linspace(-0.5,0.5, 100)[np.newaxis,:, np.newaxis]
+                    mean_wavel = np.mean(wavel)
+                    wl_matrix = a_grid * (used_wavel[np.newaxis,np.newaxis,:] - mean_wavel) + mean_wavel + b_grid
+                    template = transm_interp(wl_matrix) - 1
+                    template_std = np.mean(np.std(template, axis=-1))
+
+                    #Check if there are enough telluric features in this wavelength range
+                    if template_std>0.01:
```
Comment (tomasstolker) [inline pycrires/pipeline.py:3786]: And this (threshold for enough telluric features) is potentially another parameter that could be set by the user? From the 3 mentioned, maybe include 1 or 2 (that you think are most important) as parameters for the function.

**Review by tomasstolker — CHANGES_REQUESTED** [review]
Hooray the first PR for `pycrires` 🎉! Thanks a lot @ricolandman, this looks great! Just tested and it runs well 👍. I have added a few minor comments. I am in particular wondering if we should add a few parameters to the function with the default values as are currently used. You may know better if this is needed. I also liked the 2D plots that you showed at some point. Should we add an option to create those? Finally, could you install `pycodestyle` and run it on `pipeline.py` to fix the minor style issues (only things that are easy to fix)? Nice work!

File: pycrires/pipeline.py [hunk 7]
```diff
@@ -3726,108 +3726,90 @@ def correct_wavelengths(self, nod_ab: str = "A") -> None:
...
+                    
+                    print(
+                        f"   - {spec_name} -> lambda = {opt_b:.4f} "
+                        f"+ {opt_a:.4f} * lambda'"
                     )
 
-                hdu_list[f"CHIP{i_det+1}.INT1"].data[spec_name + "_WL"] = (
-                    opt_result.x[0] + opt_result.x[1] * spec_data[:, 0]
-                )
+                    hdu_list[f"CHIP{i_det+1}.INT1"].data[spec_name + "_WL"] = (
```
Comment (tomasstolker) [inline pycrires/pipeline.py:3803]: Should the last part be `+ mean_wavel + opt_b`?

**Comment by tomasstolker** [comment]
I am updating the `plot_spectra` function. Looks well corrected now 😊. Only the 4th order of the 1st detector seems a bit off due to the lack of telluric features. Maybe we could add a warning that prints filename and spectral order whenever `template_std` is smaller than 0.01?

spectra_nod_A_det_1_corr_000
spectra_nod_A_det_2_corr_000
spectra_nod_A_det_3_corr_000

File: pycrires/pipeline.py (hunk 1 above)
Comment (ricolandman) [reply pycrires/pipeline.py]: Done

File: pycrires/pipeline.py (hunk 2 above)
Comment (ricolandman) [reply pycrires/pipeline.py]: Done

File: pycrires/pipeline.py (hunk 3 above)
Comment (ricolandman) [reply pycrires/pipeline.py]: I have added a parameter. With mode='nearest' the results were not really stable. Instead, I have switched to another method that is a bit more stable in getting the continuum.

File: pycrires/pipeline.py (hunk 5 above)
Comment (ricolandman) [reply pycrires/pipeline.py:3793]: I think that is less important than some of the other parameters, so I have chosen to go with others instead.

File: pycrires/pipeline.py (hunk 6 above)
Comment (ricolandman) [reply pycrires/pipeline.py]: Done!

File: pycrires/pipeline.py (hunk 7 above)
Comment (ricolandman) [reply pycrires/pipeline.py:3840]: Yes indeed, no clue why I had a  minus sign there.

**Review by tomasstolker — APPROVED** [review]

**Comment by tomasstolker** [comment]
Thanks @ricolandman! Looks great! I like the plots 😊. Good idea to create an output folder in `calib` for the wavelength correction. I have also moved the FITS files there and made a few additional minor changes. I think it is ready to be merged!

**Pull request merged**